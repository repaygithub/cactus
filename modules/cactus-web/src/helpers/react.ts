import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
