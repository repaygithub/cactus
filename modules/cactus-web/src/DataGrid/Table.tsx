import { NavigationChevronDown } from '@repay/cactus-icons'
import PropTypes from 'prop-types'
import React, { ReactElement, useContext } from 'react'
import styled from 'styled-components'

import { border } from '../helpers/theme'
import Table from '../Table/Table'
import { DataGridContext } from './helpers'
import { ColumnObject, DataColumnObject, SortOption } from './types'

export interface DataGridTableProps {
  children: React.ReactNode
  data: { [key: string]: any }[]
}

const isDataColumn = (col: any): col is DataColumnObject => {
  return col && col.hasOwnProperty('sortable')
}
const isColumn = (col: any): col is ColumnObject => {
  return col && col.hasOwnProperty('columnFn')
}

const DataGridTable: React.FC<DataGridTableProps> = (props) => {
  const { children, data } = props

  const { columns, isCardView, cardBreakpoint, fullWidth, sortOptions, onSort } = useContext(
    DataGridContext
  )

  const handleSort = (id: string, exists: boolean) => {
    if (sortOptions) {
      const { sortAscending: currentSortAscending } = sortOptions[0] || {}
      const newOptions = [{ id, sortAscending: exists ? !currentSortAscending : false }]
      onSort(newOptions)
    }
  }

  return (
    <>
      {children}
      <Table fullWidth={fullWidth} cardBreakpoint={cardBreakpoint}>
        <Table.Header>
          {[...columns.keys()].map((key) => {
            const column = columns.get(key)
            if (isDataColumn(column)) {
              let sortOpt: SortOption | undefined = undefined
              if (column.sortable && sortOptions !== undefined) {
                sortOpt = sortOptions.find((opt: SortOption): boolean => opt.id === key)
              }
              const flipChevron = sortOpt !== undefined && sortOpt.sortAscending === true
              return (
                <Table.Cell
                  className={`table-cell ${flipChevron ? 'flip-chevron' : ''}`}
                  key={key}
                  aria-sort={
                    column.sortable
                      ? sortOpt
                        ? sortOpt.sortAscending
                          ? 'ascending'
                          : 'descending'
                        : 'none'
                      : undefined
                  }
                  {...column.cellProps}
                >
                  {column.sortable && sortOptions !== undefined && !isCardView ? (
                    <HeaderButton onClick={(): void => handleSort(key, sortOpt !== undefined)}>
                      <TextWrapper>{column.title}</TextWrapper>
                      <IconWrapper aria-hidden>
                        {sortOpt !== undefined && <NavigationChevronDown />}
                      </IconWrapper>
                    </HeaderButton>
                  ) : (
                    column.title
                  )}
                </Table.Cell>
              )
            } else if (isColumn(column)) {
              return (
                <Table.Cell key={`col-${key}`} {...column.cellProps}>
                  {column.title || ''}
                </Table.Cell>
              )
            }
          })}
        </Table.Header>
        <Table.Body>
          {data.map(
            (datum, datumIndex): ReactElement => (
              <Table.Row key={`row-${datumIndex}`}>
                {[...columns.keys()].map((key, keyIndex) => {
                  const column = columns.get(key)
                  if (isDataColumn(column)) {
                    const AsComponent = column.asComponent
                    return (
                      <Table.Cell key={`cell-${datumIndex}-${keyIndex}`} {...column.cellProps}>
                        {AsComponent ? <AsComponent value={datum[key]} /> : datum[key]}
                      </Table.Cell>
                    )
                  } else if (isColumn(column)) {
                    return (
                      <Table.Cell key={`cell-${datumIndex}-${keyIndex}`} {...column.cellProps}>
                        {column && column.columnFn(datum)}
                      </Table.Cell>
                    )
                  }
                })}
              </Table.Row>
            )
          )}
        </Table.Body>
      </Table>
    </>
  )
}

const IconWrapper = styled.div`
  flex-shrink: 1;
`

const TextWrapper = styled.div`
  flex-grow: 1;
`

const shapeMap = {
  square: 'border-radius: 1px;',
  intermediate: 'border-radius: 8px;',
  round: 'border-radius: 20px;',
}

const HeaderButton = styled.button`
  display: flex;
  align-items: center;
  background: none;
  border: none;
  color: inherit;
  font-size: inherit;
  text-transform: inherit;
  font-weight: inherit;
  padding: 0;
  cursor: pointer;
  position: relative;
  overflow: visible;

  &:focus {
    outline: none;

    &:after {
      content: '';
      box-sizing: border-box;
      position: absolute;
      height: calc(100% + 8px);
      width: calc(100% + 16px);
      bottom: -4px;
      right: -8px;
      border: ${(p) => border(p.theme, 'lightContrast')};
      ${(p) => shapeMap[p.theme.shape]}
    }
  }
`

DataGridTable.propTypes = {
  children: PropTypes.node.isRequired,
  data: PropTypes.arrayOf(PropTypes.object.isRequired).isRequired,
}

DataGridTable.displayName = 'Table'

export default DataGridTable
