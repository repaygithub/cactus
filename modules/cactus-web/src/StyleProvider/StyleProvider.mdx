---
name: StyleProvider
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import StyleProvider from './StyleProvider'
import cactusTheme from '@repay/cactus-theme'

# StyleProvider

The `StyleProvider` is a wrapper for `styled-components`' `ThemeProvider` component which gives you the option to include some global styles.

# Best practices

- Make sure that any components from `@repay/cactus-web` are children of `StyleProvider`. This will ensure that each component has the correct styles.

## Basic usage

```jsx
import React from 'react'
import { StyleProvider } from '@repay/cactus-web'
import cactusTheme from '@repay/cactus-theme'

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      // Include routes/rest of application here
    </StyleProvider>
  )
}
```

# Breakpoints

The `StyleProvider` has a media queries scale which is based on the provided breaking points and is added to the given theme. The media queries have screen sizes small, medium, large, and extraLarge. The set media queries will be larger than the default, for a mobile first approach.

Props can also vary depending on the current media query. This is done by setting a prop to an array, where the array holds different values for each media query size. The first element of the array maps to the default screen size, and each element after maps to the set media queries from smallest to largest. This is used below in the example code to set the background color.

For more information on Breakpoints and Media Queries visit: https://styled-system.com/theme-specification#breakpoints

## Media Queries Usage

```jsx
import React from 'react'
import { StyleProvider, Box } from '.@repay/cactus-web'
import styled from 'styled-components'

const BreakpointBox = styled(Box)`
  border-radius: 15%;

  ${p => p.theme.mediaQueries.small} {
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extraLarge} {
    border-radius: 50%;
  }
`
const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <BreakpointBox backgroundColor={backgroundColors} width="100px" height="100px" />
    </StyleProvider>
  )
}
```

# Properties

<PropsTable of={StyleProvider} staticProp="StyleProvider" />
