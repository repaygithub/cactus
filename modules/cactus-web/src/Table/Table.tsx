import { border } from '../helpers/theme'
import { Shape } from '@repay/cactus-theme'
import PropTypes from 'prop-types'
import React, { createContext, FunctionComponent, useContext } from 'react'
import styled, { StyledComponentType } from 'styled-components'

type cellAlignment = 'center' | 'right' | 'left'

type cellType = 'th' | 'td'

type BorderCorner = 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right'

interface TableContextProps {
  cellType: cellType
}

interface TableProps
  extends React.DetailedHTMLProps<React.TableHTMLAttributes<HTMLTableElement>, HTMLTableElement> {
  className?: string
  fullWidth?: boolean
}

interface TableHeaderProps
  extends React.DetailedHTMLProps<
    React.TableHTMLAttributes<HTMLTableSectionElement>,
    HTMLTableSectionElement
  > {
  className?: string
}
interface TableCellProps
  extends React.DetailedHTMLProps<
    React.TableHTMLAttributes<HTMLTableDataCellElement & HTMLTableHeaderCellElement>,
    HTMLTableDataCellElement & HTMLTableHeaderCellElement
  > {
  className?: string
  align?: cellAlignment
}

interface TableRowProps
  extends React.DetailedHTMLProps<
    React.TableHTMLAttributes<HTMLTableRowElement>,
    HTMLTableRowElement
  > {
  className?: string
}
interface TableBodyProps
  extends React.DetailedHTMLProps<
    React.TableHTMLAttributes<HTMLTableSectionElement>,
    HTMLTableSectionElement
  > {
  className?: string
}

const TableContext = createContext<TableContextProps>({
  cellType: 'td',
})

const shapeMap = {
  square: '1px',
  intermediate: '4px',
  round: '8px',
}

const getShape = (shape: Shape, location: BorderCorner) =>
  `border-${location}-radius: ${shapeMap[shape]}`

const Wrapper = styled.div<TableProps>`
  max-width: 100%;
  overflow-x: auto;
  margin: 0px 16px;
  ${(p) => p.fullWidth && 'min-width: calc(100% - 32px)'};
`
const TableBase: FunctionComponent<TableProps> = (props) => {
  const { children, className, fullWidth } = props

  return (
    <Wrapper fullWidth={fullWidth}>
      <table className={className}>{children}</table>
    </Wrapper>
  )
}

const TableCellBase: FunctionComponent<TableCellProps> = (props) => {
  const { children, className, align } = props
  const { cellType: Type } = useContext(TableContext)

  return <Type className={className}>{children}</Type>
}

const TableHeaderBase: FunctionComponent<TableHeaderProps> = (props) => {
  const { children, className } = props

  return (
    <TableContext.Provider value={{ cellType: 'th' }}>
      <thead className={className}>
        <Table.Row>{children}</Table.Row>
      </thead>
    </TableContext.Provider>
  )
}

const TableRowBase: FunctionComponent<TableRowProps> = (props) => {
  const { children, className } = props

  return (
    <tr className={className} tabIndex={0}>
      {children}
    </tr>
  )
}

const TableBodyBase: FunctionComponent<TableBodyProps> = (props) => {
  const { children, className } = props
  return (
    <TableContext.Provider value={{ cellType: 'td' }}>
      <tbody className={className}>{children}</tbody>
    </TableContext.Provider>
  )
}

export const TableHeader = styled(TableHeaderBase)`
  // Lots of specificity to override the first/last row border colors.
  &&&&& th,
  &&&&& td {
    text-transform: uppercase;
    border: ${(p) => border(p.theme, 'base')};
    ${(p) => p.theme.colorStyles.base};
  }
`

export const TableBody = styled(TableBodyBase)``
export const TableCell = styled(TableCellBase)`
  text-align: ${(p) => p.align || 'left'};
  padding: 16px;
  min-width: calc(160px * 0.7125);

  ${(p) =>
    p.theme.mediaQueries &&
    `${p.theme.mediaQueries.large} {
    min-width: calc(160px * 0.875);
  }`}

  ${(p) =>
    p.theme.mediaQueries &&
    `${p.theme.mediaQueries.extraLarge} {
    min-width: 160px;
  }`}
`

const TableRow = styled(TableRowBase)``

type TableComponentType = StyledComponentType<TableProps> & {
  Header: StyledComponentType<TableHeaderProps>
  Cell: StyledComponentType<TableCellProps>
  Row: StyledComponentType<TableRowProps>
  Body: StyledComponentType<TableBodyProps>
}

const Table = styled(TableBase)`
  ${(p) => p.fullWidth && 'min-width: 100%'};
  border-spacing: 0;
  color: ${(p) => p.theme.colors.darkestContrast};
  ${(p) => p.theme.textStyles.small};

  th {
    font-weight: bold;
  }

  td,
  th {
    background-color: ${(p) => p.theme.colors.white};
    border-top: ${(p) => border(p.theme, 'transparent')};
    border-bottom: ${(p) => border(p.theme, 'transparent')};
    :first-child {
      border-left: ${(p) => border(p.theme, 'lightContrast')};
    }
    :last-child {
      border-right: ${(p) => border(p.theme, 'lightContrast')};
    }
  }

  tr:nth-of-type(even) {
    td,
    th {
      background-color: ${(p) => p.theme.colors.lightContrast};
    }
  }

  &&& tr:hover {
    cursor: pointer;
    th,
    td {
      border-color: ${(p) => p.theme.colors.callToAction};
    }
  }

  &&& tr:focus {
    outline: 0;
    td,
    th {
      background-color: ${(p) => p.theme.colors.transparentCTA};
      border-color: ${(p) => p.theme.colors.callToAction};
    }
  }

  // first row
  & > tr:first-of-type,
  tbody > tr:first-child,
  thead > tr:first-child {
    th,
    td {
      border-top-color: ${(p) => p.theme.colors.lightContrast};
      :first-child {
        ${(p) => getShape(p.theme.shape, 'top-left')};
      }
      :last-child {
        ${(p) => getShape(p.theme.shape, 'top-right')};
      }
    }
  }

  // first row false positives
  && > thead + tr:first-of-type,
  && > thead + tbody > tr:first-child {
    th,
    td {
      border-top-color: transparent;
      border-top-left-radius: 0px;
      border-top-right-radius: 0px;
    }
  }

  // last row
  &,
  tbody:last-child,
  thead:last-child,
  tfoot:last-child {
    & > tr:last-child {
      th,
      td {
        border-bottom-color: ${(p) => p.theme.colors.lightContrast};
        :first-child {
          ${(p) => getShape(p.theme.shape, 'bottom-left')};
        }
        :last-child {
          ${(p) => getShape(p.theme.shape, 'bottom-right')};
        }
      }
    }
  }
` as TableComponentType

Table.propTypes = {
  fullWidth: PropTypes.bool,
}

Table.Header = TableHeader
Table.Cell = TableCell
Table.Row = TableRow
Table.Body = TableBody

export default Table
