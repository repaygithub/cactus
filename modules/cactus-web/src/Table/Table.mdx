---
name: Table
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import Table from './Table'
import { livePreviewStyle } from '../helpers/constants'
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'

# Table

Essentially just a table with styling to match the theme. Any valid HTML table layout should work, with a couple of exceptions:

- `rowSpan` is not really supported (but `colSpan` _is_ supported)
- Using multiple `tbody` elements is supported but because rows have odd-even styling, a `tbody` with an odd number of rows followed by another `tbody` will put two odd-styled rows next to each other.

## Basic Usage

The `Table` component has a few helpful sub-components:
- `Table.Cell` - a `td` element by default, but `as="th"` can be used to change that. It also has an `align` prop which directly translates to the CSS `text-align` property.
- `Table.Header` - a `thead` with a single row; because the row is included in the element, the direct children should be table cells. Within the `Header` the default cell type switches from `td` to `th`, but can still be overridden using `<Table.Cell as="td">`.
- `Table.Row` - a `tr` element.
- `Table.Body` - a `tbody` element.

### Scroll Wrapper

By default the table is rendered in a div with `overflow-x: auto`.
In most use cases this makes fitting a table into a site layout easier, regardless of the table's contents;
however there are some cases where you may not want the wrapper.
In particular, putting a table in a flex or grid container may result in strange sizing behavior
unless you pass `width: 100%` to match the table size to the wrapper;
or for greatest flexibility you can use the `noScrollWrapper` prop to remove the wrapper entirely.

```jsx
<Flex>
  <Flex.Item flex="1 1 400px" as={Table} noScrollWrapper>
    <Table.Header>
    ...
  </Flex.Item>
</Flex>
```

### Card Carousel

In addition to the default styling, there is another style of table that is more suitable for mobile devices. In this mode each row is rendered as a "card", where the cells are rendered as rows within the card. The cards are arranged horizontally and can be scrolled (or swiped) side-to-side, even if the scroll wrapper is removed.

There are two ways of controlling card carousel mode: using the `variant` prop, or the `cardBreakpoint` prop. `variant` has the highest precedence: if passed as `variant="table"`, the card carousel will never be used; if passed as `variant="card"`, it will always be used. The `cardBreakpoint` prop can be used to transition between table mode and card mode at one of the theme's CSS media queries, but it will only work if the app is using a `ScreenSizeProvider`. For example, the following table would be in card mode for "tiny" or "small" screen sizes, and table mode for anything larger:

```jsx
<ScreenSizeProvider>
  <Table cardBreakpoint="small">
    ...
  </Table>
</ScreenSizeProvider>
```

If a `Table.Header` is used, card mode will also render the headers in each cell/row so the user doesn't have to remember which header lines up with which cell.

### Sticky Column

For the default styling, there is the possibility to set the right most column right to be sticky. This column stays in place as the user scrolls through the table. For enabling this feature, the `Table` component allows a `sticky` prop that currently accepts `none` or `right` values.

```jsx
<ScreenSizeProvider>
  <Table sticky="right">
    ...
  </Table>
</ScreenSizeProvider>
```

### Try it out

export const code = `
<Table>
  <Table.Header>
    <Table.Cell colSpan="2">Spanned Header</Table.Cell>
    <Table.Cell>Another Column</Table.Cell>
  </Table.Header>
  <Table.Row>
    <Table.Cell as="th">SubHeader</Table.Cell>
    <Table.Cell>SubValue</Table.Cell>
    <Table.Cell align="right">Another Value</Table.Cell>
  </Table.Row>
  <Table.Row>
    <Table.Cell align="center" colSpan="2">Spanned SubHeader</Table.Cell>
    <Table.Cell align="right">That One Value</Table.Cell>
  </Table.Row>
</Table>
`

<LiveProvider code={code} scope={{ Table }}>
  <LiveEditor style={livePreviewStyle} />
  <LiveError />
  <LivePreview />
</LiveProvider>

## Properties

<PropsTable of={Table} />
