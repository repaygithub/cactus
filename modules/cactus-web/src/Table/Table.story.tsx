import React from 'react'

import { DescriptiveEnvelope } from '@repay/cactus-icons'
import { storiesOf } from '@storybook/react'
import Table from './Table'

storiesOf('Table', module).add('Header', () => (
  <Table>
    <Table.Header>
      <Table.Cell cellType="th" align="center">
        Header
      </Table.Cell>
      <Table.Cell cellType="th" align="center">
        Header
      </Table.Cell>
      <Table.Cell cellType="th" align="center">
        Header
      </Table.Cell>
    </Table.Header>
  </Table>
))

const arr = [1, 2, 3, 4, 5, 6, 7]

const amount = [
  '$ 123.00',
  '$ 123.00',
  '$ 123.00',
  '$ 123.00',
  '$ 123.000.00',
  '$ 23.00',
  '$ 3000.00',
  '$ 123.231.00',
]

storiesOf('Table', module).add('Row', () => (
  <Table>
    <Table.Body>
      <Table.Row>
        <>
          {arr.map((e, i) => (
            <Table.Cell cellType="td" align="left" key={i}>
              Cell {e}
            </Table.Cell>
          ))}
          <Table.Cell cellType="td" align="center">
            Label
            <DescriptiveEnvelope />
          </Table.Cell>
        </>
      </Table.Row>
      <Table.Row>
        {amount.map((e, i) => (
          <Table.Cell cellType="td" align="right" key={i}>
            {e}
          </Table.Cell>
        ))}
      </Table.Row>
      <Table.Row>
        <>
          <Table.Cell cellType="td" align="center">
            <DescriptiveEnvelope />
          </Table.Cell>
          {arr.map((e, i) => (
            <Table.Cell cellType="td" align="left" key={i}>
              Cell {e}
            </Table.Cell>
          ))}
        </>
      </Table.Row>
    </Table.Body>
  </Table>
))
