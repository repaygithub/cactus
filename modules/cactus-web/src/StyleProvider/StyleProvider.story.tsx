import React from 'react'

import { storiesOf } from '@storybook/react'
import Box from '../Box/Box'
import cactusTheme from '@repay/cactus-theme'
import Grid from '../Grid/Grid'
import styled from 'styled-components'
import StyleProvider from './StyleProvider'

const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const BreakpointBox = styled(Box)`
  max-width: 320px;
  width: 50px;
  height: 50px;
  border-radius: 15%

  ${p => p.theme.mediaQueries.small} {
    max-width: 708px;
    width: 75px;
    height 75px;
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    max-width: 964px;
    width: 100px;
    height: 100px;
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    max-width: 1140px;
    width: 150px;
    height: 150px;
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extralarge} {
    max-width: 1380px;
    width: 200px;
    height: 200px;
    border-radius: 50%;
  }
`

storiesOf('Style Provider', module).add('Component Adjusments based on Media Queries', () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <Grid justify="center">
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
      </Grid>
    </StyleProvider>
  )
})
