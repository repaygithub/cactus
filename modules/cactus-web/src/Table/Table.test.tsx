import { cleanup, render } from '@testing-library/react'
import * as React from 'react'

import { StyleProvider } from '../StyleProvider/StyleProvider'
import Table from './Table'

afterEach(cleanup)

describe('component: Table', (): void => {
  test('regular table', (): void => {
    const { container } = render(
      <StyleProvider>
        <Table>
          <Table.Header>
            <Table.Cell>Header Cell</Table.Cell>
          </Table.Header>
          <Table.Body>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
          </Table.Body>
        </Table>
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('table with no header', (): void => {
    const { container } = render(
      <StyleProvider>
        <Table>
          <Table.Body>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
          </Table.Body>
        </Table>
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('table with caption', (): void => {
    const { container } = render(
      <StyleProvider>
        <Table>
          <caption>I'm a caption</caption>
          <Table.Header>
            <Table.Cell>Header Cell</Table.Cell>
          </Table.Header>
          <Table.Body>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
          </Table.Body>
        </Table>
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('card carousel', (): void => {
    const { container } = render(
      <StyleProvider>
        <Table variant="card">
          <Table.Header>
            <Table.Cell>Header Cell</Table.Cell>
          </Table.Header>
          <Table.Body>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
          </Table.Body>
        </Table>
      </StyleProvider>
    )

    expect(container).toMatchSnapshot()
  })
  test('ignores width on card', (): void => {
    const { getByTestId } = render(
      <StyleProvider>
        <Table variant="card">
          <Table.Header>
            <Table.Cell>Header Cell</Table.Cell>
          </Table.Header>
          <Table.Body>
            <Table.Row>
              <Table.Cell width="230px" data-testid="cell">
                Data cell
              </Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Data cell</Table.Cell>
            </Table.Row>
          </Table.Body>
        </Table>
      </StyleProvider>
    )
    const singleCell = getByTestId('cell')
    expect(singleCell).toHaveStyle('width: 272px')
  })
})
