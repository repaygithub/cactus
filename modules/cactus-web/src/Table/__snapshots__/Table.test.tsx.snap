// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: Table snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0 16px;
}

<div>
  <table
    class="c0"
  />
</div>
`;
